### PR TITLE
Adding S3 Versioning functionality to code

### PR DESCRIPTION
**Pull request by homernajafi**
This will add Setter, Getter functions for S3 Versioning.

### CODE REVIEW

File: util/pkg/vfs/s3fs.go [hunk 1]
```diff
@@ -61,6 +61,14 @@ func newS3Path(s3Context *S3Context, bucket string, key string) *S3Path {
 }
 
 func (p *S3Path) Path() string {
+	response,_ := p.GetVersioning()
+	if response != true {
+		fmt.Println("S3 Bucket Versioning is not set. Let's set it")
+		p.SetVersioning("Enabled")
```
Comment (chrislovecnm) [inline util/pkg/vfs/s3fs.go]: I do not think we can enable versioning for everyone.  Let's just log that it is not on.

File: util/pkg/vfs/s3fs.go [hunk 2]
```diff
@@ -61,6 +61,14 @@ func newS3Path(s3Context *S3Context, bucket string, key string) *S3Path {
 }
 
 func (p *S3Path) Path() string {
+	response,_ := p.GetVersioning()
+	if response != true {
+		fmt.Println("S3 Bucket Versioning is not set. Let's set it")
```
Comment (chrislovecnm) [inline util/pkg/vfs/s3fs.go]: use glog here.  Log a warning.

File: util/pkg/vfs/s3fs.go [hunk 3]
```diff
@@ -61,6 +61,14 @@ func newS3Path(s3Context *S3Context, bucket string, key string) *S3Path {
...
+		p.SetVersioning("Enabled")
+	} else {
+		fmt.Println("S3 Bucket Versioning is set already.")
```
Comment (chrislovecnm) [inline util/pkg/vfs/s3fs.go]: Just pull the else, unless you want to put in a glog that is like V(8).

File: util/pkg/vfs/s3fs.go [hunk 4]
```diff
@@ -155,6 +163,50 @@ func (p *S3Path) WriteFile(data io.ReadSeeker, aclObj ACL) error {
...
+
+	request := &s3.GetBucketVersioningInput{}
+	request.SetBucket(p.bucket)
+
+	response, err := client.GetBucketVersioning(request)
+	if err != nil {
+		return false, fmt.Errorf("error getting s3 bucket versioning: %v", p, err)
+	}
+
+	if strings.Contains(response.String(),"Status: \"Enabled\"") {
```
Comment (chrislovecnm) [inline util/pkg/vfs/s3fs.go]: Is there a value in the AWS APU for this?  I would rather not test a string response.

File: util/pkg/vfs/s3fs.go [hunk 5]
```diff
@@ -155,6 +163,50 @@ func (p *S3Path) WriteFile(data io.ReadSeeker, aclObj ACL) error {
...
+		return true, nil
+	} else {
+		return false, nil
+	}
+}
+
+func (p *S3Path) SetVersioning(status string) ( error) {
```
Comment (chrislovecnm) [inline util/pkg/vfs/s3fs.go]: Please add godocs here. We will need to expose this through the Interface, but we need to figure out how to do it first.  Let's just log first.

File: util/pkg/vfs/s3fs.go [hunk 6]
```diff
@@ -155,6 +163,50 @@ func (p *S3Path) WriteFile(data io.ReadSeeker, aclObj ACL) error {
...
+	client, err := p.client()
+	if err != nil {
+		return err
+	}
+
+	requestConfig := &s3.VersioningConfiguration{}
```
Comment (chrislovecnm) [inline util/pkg/vfs/s3fs.go]: Please construct as an inline struct.  You can inline all of the struct members.

```go
requestConfig := &s3.VersioningConfiguration{
  SetStatus: status,
}
```

Comment (homernajafi) [reply util/pkg/vfs/s3fs.go]: It doesn't work for me somehow. It says "SetStatus not found"

File: util/pkg/vfs/s3fs.go (hunk 5 above)
Comment (homernajafi) [reply util/pkg/vfs/s3fs.go]: Adding to interface means to implement it for other storage types too. This code is only for S3. Do you want me to add empty functions in other types too?

File: util/pkg/vfs/s3fs.go (hunk 4 above)
Comment (homernajafi) [reply util/pkg/vfs/s3fs.go]: This was the only way I could find. Return value doesn't have a status flag.

File: util/pkg/vfs/s3fs.go (hunk 3 above)
Comment (homernajafi) [reply util/pkg/vfs/s3fs.go]: done

File: util/pkg/vfs/s3fs.go (hunk 2 above)
Comment (homernajafi) [reply util/pkg/vfs/s3fs.go]: done

File: util/pkg/vfs/s3fs.go (hunk 1 above)
Comment (homernajafi) [reply util/pkg/vfs/s3fs.go]: done

**Pull request closed without merging**